src/Cassandra/Mapping/Statements/StatementFactory.cs

```diff
@@ -78,11 +80,12 @@ public async Task<Statement> GetStatementAsync(ISession session, Cql cql, bool?
             {
                 // The exception was caused from awaiting upon a Task that was previously cached
                 // It's possible that the schema or topology changed making this query preparation to succeed
-                // in a new attemp
-                prepareTask = session.PrepareAsync(query);
+                // in a new attempt
+                prepareTask = _statementCache.CompareAndUpdate(
```

Comment:
I think we can make `CompareAndUpdate()` more generic with 2 params: `key` and `valueFactory`, then it can be used in the following way:

```csharp
// If there is a new prepare task from other thread, using the new one without making new requests
cache.CompareAndUpdate(key, (k, v) => v != prepareTask ? v : Task.Run(() => session.PrepareAsync(query)));
```
wdyt?

Reply:
yeah that's a good idea, changed it!